integrations/access/opsgenie/app.go

```diff
@@ -17,15 +17,573 @@ limitations under the License.
+		// If plugin data is empty or missing alertID, we cannot do anything.
+		if existing == nil {
+			return PluginData{}, false
+		}
+		if alertID = existing.AlertID; alertID == "" {
+			return PluginData{}, false
+		}
+
+		// If resolution field is not empty then we already resolved the alert before. In this case we just quit.
+		if existing.RequestData.Resolution.Tag != Unresolved {
+			return PluginData{}, false
+		}
+
+		// Mark alert as resolved.
+		pluginData := *existing
+		pluginData.Resolution = resolution
+		return pluginData, true
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	if !ok {
+		logger.Get(ctx).Debug("Failed to resolve the alert: plugin data is missing")
+		return nil
+	}
+
+	ctx, log := logger.WithField(ctx, "opsgenie_alert_id", alertID)
+	if err := a.opsgenie.ResolveAlert(ctx, alertID, resolution); err != nil {
+		return trace.Wrap(err)
+	}
+	log.Info("Successfully resolved the alert")
+
+	return nil
+}
+
+// modifyPluginData performs a compare-and-swap update of access request's plugin data.
+// Callback function parameter is nil if plugin data hasn't been created yet.
+// Otherwise, callback function parameter is a pointer to current plugin data contents.
+// Callback function return value is an updated plugin data contents plus the boolean flag
+// indicating whether it should be written or not.
+// Note that callback function fn might be called more than once due to retry mechanism baked in
+// so make sure that the function is "pure" i.e. it doesn't interact with the outside world:
+// it doesn't perform any sort of I/O operations so even things like Go channels must be avoided.
+// Indeed, this limitation is not that ultimate at least if you know what you're doing.
+func (a *App) modifyPluginData(ctx context.Context, reqID string, fn func(data *PluginData) (PluginData, bool)) (bool, error) {
+	backoff := backoff.NewDecorr(modifyPluginDataBackoffBase, modifyPluginDataBackoffMax, clockwork.NewRealClock())
+	for {
+		oldData, err := a.getPluginData(ctx, reqID)
+		if err != nil && !trace.IsNotFound(err) {
+			return false, trace.Wrap(err)
+		}
+		newData, ok := fn(oldData)
+		if !ok {
+			return false, nil
+		}
+		var expectData PluginData
+		if oldData != nil {
+			expectData = *oldData
+		}
+		err = trace.Wrap(a.updatePluginData(ctx, reqID, newData, expectData))
+		if err == nil {
+			return true, nil
+		}
+		if !trace.IsCompareFailed(err) {
+			return false, trace.Wrap(err)
+		}
+		if err := backoff.Do(ctx); err != nil {
+			return false, trace.Wrap(err)
+		}
+	}
+}
+
+// getPluginData loads a plugin data for a given access request. It returns nil if it's not found.
+func (a *App) getPluginData(ctx context.Context, reqID string) (*PluginData, error) {
+	dataMaps, err := a.teleport.GetPluginData(ctx, types.PluginDataFilter{
+		Kind:     types.KindAccessRequest,
+		Resource: reqID,
+		Plugin:   pluginName,
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	if len(dataMaps) == 0 {
```

Comment:
Is there any risks multiple dataMaps may be returned ?

Reply:
There shouldn't be. I can't think of a way more than 1 would exist per reqID